packages/xstate-solid/src/createImmutable.ts

```diff
@@ -0,0 +1,114 @@
+  let proto;
+  return (
+    obj != null &&
+    typeof obj === 'object' &&
+    (!(proto = Object.getPrototypeOf(obj)) ||
+      proto === Object.prototype ||
+      Array.isArray(obj))
+  );
+}
+
+const resolvePath = (path: any[], obj = {}) =>
+  path.reduce((prev, curr) => prev?.[curr], obj);
+
+const updateStore = <Path extends unknown[]>(
+  nextStore: Store<any>,
+  prevStore: Store<any>,
+  set: (...args: [...Path, any?, any?]) => void,
+  store: Store<any>
+) => {
+  const valueRefs = new WeakMap<any, unknown>();
+  const diff = <CompareValue extends unknown>(
+    next: CompareValue,
+    prev: CompareValue,
+    path: Path
+  ) => {
+    if (prev === next) {
+      return;
+    }
+
+    // Use reference if it has already been used circular reference loops
+    if (valueRefs.has(next)) {
+      set(...path, valueRefs.get(next));
+      return;
+    }
+
+    if (!isWrappable(next) || !isWrappable(prev)) {
+      // toString cannot be set in solid stores
+      if (path[path.length - 1] !== 'toString') {
+        set(...path, () => next);
+      }
+      return;
+    }
+
+    // next is either an object or array, save reference to prevent diffing
+    // the same object twice
+    valueRefs.set(next, resolvePath(path, store));
+
+    // Diff and update array or object
+    if (Array.isArray(next) && Array.isArray(prev)) {
+      const newIndices = next.length - prev.length;
+      const smallestSize = Math.min(prev.length, next.length);
+      const largestSize = Math.max(next.length, prev.length);
+
+      // Diff array
+      for (let start = 0, end = smallestSize - 1; start < end; start++, end--) {
+        diff(next[start], prev[start], [...path, start] as Path);
+        diff(next[end], prev[end], [...path, end] as Path);
+      }
+
+      // Update new or now undefined indexes
+      if (newIndices !== 0) {
+        for (let newEnd = smallestSize; newEnd <= largestSize - 1; newEnd++) {
+          set(...path, newEnd, next[newEnd]);
+        }
+        if (prev.length > next.length) {
+          set(...path, 'length', next.length);
+        }
+      }
+    } else {
+      // Update new values
+      const targetKeys = Object.keys(next) as Array<keyof CompareValue>;
+      for (let i = 0, len = targetKeys.length; i < len; i++) {
+        diff(next[targetKeys[i]!], prev[targetKeys[i]!], [
+          ...path,
+          targetKeys[i]
+        ] as Path);
+      }
+
+      // Remove previous keys that are now undefined
+      const previousKeys = Object.keys(prev) as Array<keyof CompareValue>;
+      for (let i = 0, len = previousKeys.length; i < len; i++) {
+        if (next[previousKeys[i]!] === undefined) {
+          set(...path, previousKeys[i]!, undefined);
+        }
+      }
+    }
+  };
+  diff(nextStore, prevStore, [] as any);
+};
+
+/**
+ * Based on Ryan Carniato's createImmutable prototype
+ * Clones the initial value and diffs updates
+ */
+export function createImmutable<T extends object | []>(
+  init: T
```

Comment:
There are some minor type vs runtime inconsistencies related to this. It's said here that `init` is **not** a primitive (as it `extends object`). The code later, in `deepClone`, `isWrappable`, `resolvePath` handles `undefined` and `null` (and potentially more) though. 

Should this constraint here be relaxed? or is using `createImmutable` with primitives not possible? If it's not possible then perhaps we could drop the code related to handling primitives. I know that this other code might assume more generic context - but we don't handle those cases then we don't need to assume that more general context ;p (aka shave bytes when you can)

Reply:
I updated the init arg to only accept objects. Because `deepClone`, `isWrappable`, `resolvePath`, and `diff` are all called while recursively checking a store, they need to handle the top level object as well as its children. `createImmutable` will only handle objects passed in from one of the hooks, all the helper functions check the entire tree of the object

Reply:
Ah, I see this now - thanks for the clarification 👍